builtin/v9/migration/miner.go

```diff
@@ -165,3 +179,262 @@ func (m minerMigrator) migrateState(ctx context.Context, store cbor.IpldStore, i
+		var unsealedCid *cid.Cid
+		var pieces []abi.PieceInfo
+		for _, dealID := range info.Info.DealIDs {
+			deal, err := m.proposals.GetDealProposal(dealID)
+			if err != nil {
+				// Possible for the proposal to be missing if it's expired (but the deal is still in a precommit that's yet to be cleaned up)
+				// Just continue in this case, the sector is unProveCommitable anyway, will just fail later
+				if exitcode.Unwrap(err, exitcode.ErrIllegalState) != exitcode.ErrNotFound {
+					return xerrors.Errorf("error getting deal proposal for sector: %d: %w", info.Info.SectorNumber, err)
+				}
+
+				continue
+			}
+
+			pieces = append(pieces, abi.PieceInfo{
+				PieceCID: deal.PieceCID,
+				Size:     deal.PieceSize,
+			})
+		}
+
+		if len(pieces) != 0 {
+			commd, err := commp.GenerateUnsealedCID(info.Info.SealProof, pieces)
+			if err != nil {
+				return xerrors.Errorf("failed to generate unsealed CID: %w", err)
+			}
+
+			unsealedCid = &commd
+		}
+
+		err = newPrecommitOnChainInfos.Put(miner9.SectorKey(info.Info.SectorNumber), &miner9.SectorPreCommitOnChainInfo{
+			Info: miner9.SectorPreCommitInfo{
+				SealProof:     info.Info.SealProof,
+				SectorNumber:  info.Info.SectorNumber,
+				SealedCID:     info.Info.SealedCID,
+				SealRandEpoch: info.Info.SealRandEpoch,
+				DealIDs:       info.Info.DealIDs,
+				Expiration:    info.Info.Expiration,
+				UnsealedCid:   unsealedCid,
+			},
+			PreCommitDeposit: info.PreCommitDeposit,
+			PreCommitEpoch:   info.PreCommitEpoch,
+		})
+
+		if err != nil {
+			return xerrors.Errorf("failed to write new precommitinfo: %w", err)
+		}
+
+		return nil
+	})
+
+	if err != nil {
+		return cid.Undef, xerrors.Errorf("failed to iterate over precommitinfos: %w", err)
+	}
+
+	newPrecommits, err := newPrecommitOnChainInfos.Root()
+	if err != nil {
+		return cid.Undef, xerrors.Errorf("failed to flush new precommits: %w", err)
+	}
+
+	return newPrecommits, nil
+}
+
+func migrateSectorsWithCache(ctx context.Context, store adt8.Store, cache MigrationCache, minerAddr address.Address, inRoot cid.Cid) (cid.Cid, error) {
+	return cache.Load(SectorsAmtKey(inRoot), func() (cid.Cid, error) {
+		inArray, err := adt8.AsArray(store, inRoot, miner8.SectorsAmtBitwidth)
+		if err != nil {
+			return cid.Undef, xerrors.Errorf("failed to read sectors array: %w", err)
+		}
+
+		okIn, prevInRoot, err := cache.Read(MinerPrevSectorsInKey(minerAddr))
+		if err != nil {
+			return cid.Undef, xerrors.Errorf("failed to get previous inRoot from cache: %w", err)
+		}
+
+		okOut, prevOutRoot, err := cache.Read(MinerPrevSectorsOutKey(minerAddr))
+		if err != nil {
+			return cid.Undef, xerrors.Errorf("failed to get previous outRoot from cache: %w", err)
+		}
+
+		var outArray *adt9.Array
+		if okIn && okOut {
+			// we have previous work, but the AMT has changed -- diff them
+			diffs, err := amt.Diff(ctx, store, store, prevInRoot, inRoot, amt.UseTreeBitWidth(miner9.SectorsAmtBitwidth))
```

Comment:
I'm not sure if I'm reading it right but [Diff](https://github.com/filecoin-project/go-amt-ipld/blob/723fff1300c57c2f866505a7f3bd0796817e8bbf/diff.go#L42) doesn't seem to quickly check that the cids are equal and return immediately.  If this is right it is important for us to check equality quickly and set outArray to prevOutArray.  It might speed us up a lot on second run.

Reply:
Aha I'm missingt that this is inside cache.Load